File: backend/app/utils/email_service.py
import os
import requests
from typing import Dict, Any, Optional

def load_env_var(name: str, default: str = "") -> str:
    val = os.environ.get(name)
    if val is not None:
        return val
    for path in [".env", "../.env", "backend/.env", "app/.env"]:
        if os.path.exists(path):
            try:
                with open(path, "r", encoding="utf-8") as f:
                    for line in f:
                        line = line.strip()
                        if not line or line.startswith("#"):
                            continue
                        if "=" in line:
                            k, v = line.split("=", 1)
                            if k.strip() == name:
                                return v.strip().strip('"').strip("'")
            except Exception:
                pass
    return default

EMAILJS_SERVICE_ID = load_env_var("EMAILJS_SERVICE_ID", "service_fk7d9bp")
EMAILJS_TEMPLATE_ID = load_env_var("EMAILJS_TEMPLATE_ID", "template_yunwgcv")
EMAILJS_PUBLIC_KEY = load_env_var("EMAILJS_PUBLIC_KEY", "IJxjBjS0m-gQq8883")
EMAILJS_API_URL = "https://api.emailjs.com/api/v1.0/email/send"

def send_email_notification(
    to_email: str,
    client_name: str,
    contract_title: str,
    end_date: str,
    days_left: int,
    custom_message: Optional[str] = None,
    employer_name: Optional[str] = None,
    company_name: Optional[str] = None
) -> Dict[str, Any]:
    """
    Sends real email notification to the recipient using EmailJS REST API.
    """
    if not custom_message:
        custom_message = (
            f"Hello {client_name},\n\n"
            f"This is a reminder that your contract '{contract_title}' will expire on {end_date}.\n\n"
            f"There are {days_left} days left for the contract to expire.\n\n"
            f"Regards,\nAI Contract Management System"
        )
    
    subject = f"Contract Expiry Reminder: {contract_title} ({days_left} days left)"

    # Comprehensive template parameter aliases to match EmailJS template variables
    template_params = {
        "to_email": to_email,
        "email": to_email,
        "recipient_email": to_email,
        "client_email": to_email,
        "to_name": client_name,
        "client_name": client_name,
        "name": client_name,
        "contract_title": contract_title,
        "contract_name": contract_title,
        "filename": contract_title,
        "end_date": str(end_date),
        "expiry_date": str(end_date),
        "days_left": str(days_left),
        "days_until_expiry": str(days_left),
        "message": custom_message,
        "body": custom_message,
        "subject": subject,
        "employer_name": employer_name or "AI Contract Management",
        "company_name": company_name or "AI Contract Management"
    }

    payload = {
        "service_id": EMAILJS_SERVICE_ID,
        "template_id": EMAILJS_TEMPLATE_ID,
        "user_id": EMAILJS_PUBLIC_KEY,
        "template_params": template_params
    }

    headers = {
        "Content-Type": "application/json",
        "User-Agent": "AI-Contract-Management-System/1.0"
    }

    try:
        response = requests.post(EMAILJS_API_URL, json=payload, headers=headers, timeout=10)
        if response.status_code in [200, 201]:
            print(f"[EmailJS] [SUCCESS] Email dispatched successfully to {to_email} (Service: {EMAILJS_SERVICE_ID}, Template: {EMAILJS_TEMPLATE_ID})")
            return {"success": True, "status_code": response.status_code, "message": "Email sent successfully via EmailJS."}
        else:
            print(f"[EmailJS] [WARNING] EmailJS returned status {response.status_code}: {response.text}")
            return {"success": False, "status_code": response.status_code, "error": response.text}
    except Exception as e:
        print(f"[EmailJS] [ERROR] Failed to send email via EmailJS: {e}")
        return {"success": False, "error": str(e)}

File: frontend/src/utils/emailService.ts
import emailjs from '@emailjs/browser';
import { api, type Contract } from '../api/client';

export const EMAILJS_CONFIG = {
  SERVICE_ID: import.meta.env.VITE_EMAILJS_SERVICE_ID || 'service_fk7d9bp',
  TEMPLATE_ID: import.meta.env.VITE_EMAILJS_TEMPLATE_ID || 'template_yunwgcv',
  PUBLIC_KEY: import.meta.env.VITE_EMAILJS_PUBLIC_KEY || 'IJxjBjS0m-gQq8883',
};

export interface SendEmailParams {
  toEmail: string;
  clientName: string;
  contractTitle: string;
  endDate: string;
  daysLeft: number;
  message?: string;
  employerName?: string;
  companyName?: string;
}

export async function sendBrowserEmailNotification(params: SendEmailParams) {
  const {
    toEmail,
    clientName,
    contractTitle,
    endDate,
    daysLeft,
    message,
    employerName = 'AI Contract Management',
    companyName = 'AI Contract Management',
  } = params;

  const defaultMsg = message || `Hello ${clientName},\n\nThis is a reminder that your contract '${contractTitle}' will expire on ${endDate}.\n\nThere are ${daysLeft} days left for the contract to expire.\n\nRegards,\nAI Contract Management System`;

  const templateParams: Record<string, any> = {
    // Recipient & sender fields
    to_email: toEmail,
    email: toEmail,
    recipient_email: toEmail,
    client_email: toEmail,
    user_email: toEmail,
    to_name: clientName,
    client_name: clientName,
    name: clientName,
    from_name: employerName || 'AI Contract Management',
    reply_to: 'no-reply@contractmgmt.com',
    
    // Contract details
    contract_title: contractTitle,
    contract_name: contractTitle,
    filename: contractTitle,
    end_date: String(endDate),
    expiry_date: String(endDate),
    days_left: String(daysLeft),
    days_until_expiry: String(daysLeft),
    
    // Message contents & Subject
    message: defaultMsg,
    body: defaultMsg,
    subject: `Contract Expiry Reminder: ${contractTitle} (${daysLeft} days left)`,
    employer_name: employerName,
    company_name: companyName,
  };

  const response = await emailjs.send(
    EMAILJS_CONFIG.SERVICE_ID,
    EMAILJS_CONFIG.TEMPLATE_ID,
    templateParams,
    EMAILJS_CONFIG.PUBLIC_KEY
  );

  return response;
}

export async function dispatchContractExpiryNotification(contract: Contract, customMessage?: string) {
  const toEmail = contract.client_email;
  if (!toEmail || !toEmail.trim()) {
    throw new Error('This contract does not have a client email address specified.');
  }

  const contractTitle = contract.filename || `Contract #${contract.id} (${contract.employer_name} - ${contract.client_name})`;

  // 1. Send real email directly from browser via EmailJS (avoids 403 non-browser API restriction)
  const emailRes = await sendBrowserEmailNotification({
    toEmail: toEmail.trim(),
    clientName: contract.client_name,
    contractTitle,
    endDate: contract.end_date,
    daysLeft: contract.days_until_expiry,
    message: customMessage,
    employerName: contract.employer_name,
    companyName: contract.company_name,
  });

  // 2. Notify backend to update database status and record audit log
  try {
    await api.sendExpiryNotification(contract.id);
  } catch (backendErr) {
    console.warn('Backend audit log sync error:', backendErr);
  }

  return { success: true, email: toEmail, status: emailRes.status, text: emailRes.text };
}
